### DISCUSSION TITLE
資料庫預設時間問題 - 時區對但時間錯誤

### DISCUSSION THREAD

**Discussion by weih210287** [Lidemy/forum#154] [後端相關問題] [answered]
這邊在操作課程共用的 phpMyAdmin 時遇到了時間設定上的問題。

是這樣的，我在 conn.php 裡已經有設定好 `$conn->query('SET time_zone = "+8:00"');`，之後設定 current_at 的內容時預設了時間是 current_time_stamp，可是最後不管是直接在 index.php 網頁上測試還是在資料庫測試新增內容，最後顯示出的時間雖然日期是正確的今日，但時間都是從凌晨開始算，如圖，我開始新增的時間大概是 7/1 15:40 左右：

![](https://i.imgur.com/R3VqEA8.png)
<br>
而打開 current_time_stamp 時區也確實是 UTC+8 沒錯：

![](https://i.imgur.com/VHqhHwY.jpg)
<br>
查了一下資料後，我有試著在 conn.php 裡增加 `  date_default_timezone_set('Asia/Taipei');
` 不過也沒什麼用 orz
我在使用 xampp 自己的 phpMyAdmin 時日期跟時間都是正確的，只有在共用的 phpMyAdmin 才出問題，FileZilla 顯示新增檔案的時間也是正確的
想請問大家之前交作業在設定時間時也有遇過這種問題嗎 qwq？

**Comment by cwc329** [comment]
先說個不相關的，
資料庫與操作資料庫的工具是不一樣的。

我們課程用的資料庫應該是 MySQL，
然後 XAMPP 用的沒記錯的話是 MariaDB。

phpMyAdmin 是操作資料庫的一項工具，
不是資料庫。
今天你不用 phpMyAdmin，
用 DBeaver, MySQLWorkBench 甚至是 cli 直接操作 DB 也可以，
兩者是不一樣的。

這個問題應該是 MySQL or MariaDB 的設定問題，
而不是 phpMyAdmin。

**Reply by weih210287** [reply]
原來如此，這部分我真的沒有搞清楚 😅
謝謝助教，我再去研究一下

**Comment by aszx87410** [comment, marked as answer]
試試看這幾篇裡面提到的好了
https://stackoverflow.com/questions/930900/how-do-i-set-the-time-zone-of-mysql
https://www.brilliantcode.net/473/mysql-set-change-timezone/

SET @@session.time_zone = '+08:00';
SET GLOBAL time_zone = '+08:00';

**Reply by weih210287** [reply]
試了一下 `@@session.time_zone`，改完之後從留言板更新時間就正確沒問題了，謝謝老師！

**Comment by weih210287** [comment]
參考了助教和老師的解答後，不太確定自己做的到底是什麼設定然後他們又有什麼差別，之後又看到了一篇 [文章](http://bonvoyagelin.blogspot.com/2011/01/mysql.html)，才稍微了解，紀錄一下自己的小總結：

原來 MySQL 伺服器的時區和我從網頁 Client 端的時區是不太一樣的，用 `@@session.time_zone` 可以調整我在 Client 端的時區，然後如果有 SUPER 權限就可以使用 `GLOBAL` 去更改伺服器的時區，所以成功用了 `@@session.time_zone` 之後我在網頁新增的留言顯示時間就會是自己 local 的時間，但如果從 MySQL 伺服器新增的就還是會不一樣，因為他的預設時區和我本地的可能不一樣。
